documentation/writing_tests.md

```diff
@@ -0,0 +1,1534 @@
+
+Let's say we have a client that has has three main functionalities: a constructor, a method `A` that can only be accessed through the use of the instantiated client, and a method `AB` that is only valuable after `A` is called, like in the following example:
+
+```ts
+export class ClientMethodDependency {
+  public history: string[];
+
+  constructor() {
+    this.history = ["constructor"];
+  }
+
+  A(): void {
+    this.history.push("A");
+  }
+
+  AB(): void {
+    if (this.history[this.history.length - 1] !== "A") {
+      throw new Error("A should have been called");
+    }
+    this.history.push("B");
+  }
+}
+```
+
+We could write three tests, as follows:
+
+```ts
+describe("Tests for the ClientMethodDependency class", function() {
+
+  it("The ClientMethodDependency should initialize and set its history property properly", function() {
+    const client = new ClientMethodDependency();
+    assert.deepEqual(client.history, ["constructor"]);
+  });
+
+  it("The ClientMethodDependency's A method should alter the client's history", function() {
+    const client = new ClientMethodDependency();
+    client.A();
+    assert.deepEqual(client.history, ["constructor", "A"]);
+  });
+
+  it("The ClientMethodDependency's AB method should throw if it's called before A", function() {
+    const client = new ClientMethodDependency();
+    assert.throws(client.AB);
+  });
+
+  it("The ClientMethodDependency's AB method should alter the client's history after A is called", function() {
+    const client = new ClientMethodDependency();
+    client.A();
+    assert.deepEqual(client.history, ["constructor", "A"]);
+    client.AB();
+    assert.deepEqual(client.history, ["constructor", "A", "AB"]);
+  });
+
+});
+```
+
+Each test's goal is to verify the behavior of a single functionality, even though some have to rely on a more complex setup than others.
+
+If there are no function calls in a test, a single test case must compare the values of properties with similar names and similar values, for example:
+
+```ts
+import { version } from "../some/path/constants.ts";
+
+describe("testing some of the client's public properties", function() {
+  let client: Client;
+
+  beforeEach(function() {
+    recorder = record(this);
+    client = new Client();
+  });
+
+  afterEach(function () {
+    recorder.stop();
+  });
+
+  it("should have a valid version", function() {
+    assert.equal(client.version, version)
+  });
+});
+```
+
+It's valid to test more than one property from the result of a single function, like in the following example:
+
+```ts
+describe("Tests with more than one property", function() {
+  let client: Client;
+
+  beforeEach(function() {
+    recorder = record(this);
+    client = new Client();
+  });
+
+  afterEach(function () {
+    recorder.stop();
+  });
+
+  it("should test A", async function() {
+    const result = await client.A();
+    assert.ok(result.value);
+    assert.ok(result.createdAt instanceof Date);
```

Comment:
nit:  should use `assert.typeOf` or `assert.instanceOf`

https://www.chaijs.com/api/assert/#method_typeof 
https://www.chaijs.com/api/assert/#method_instanceof